File: SwapNodesinPairs.cpp
#include <iostream>
using namespace std;


struct ListNode 
{
   int val;
   ListNode *next;
   ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode *swapPairs(ListNode *head) 
    {
        ListNode *pre = NULL;
        ListNode *cur = head;
        ListNode *nextN = NULL;
        while (cur != NULL && cur->next != NULL)
        {
        	nextN = cur->next;
        	if (pre != NULL) 
        	{
        		pre->next = nextN;
        	}
        	else
        	{
        		head = nextN;
        	}
        	cur->next = nextN->next;
        	nextN->next = cur;
        	pre = cur;
        	cur = cur->next;
        }
        return head;
    }
};

int main()
{
	return 0;
}

File: JavaSolution/src/Microsoft/SwapNodesinPairs.java
package Microsoft;

/**
 * Created by xyunpeng on 2/20/16.
 */
public class SwapNodesinPairs {
    public ListNode swapPairs(ListNode head) {
        if (head == null) return null;
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode pre = dummy;
        ListNode cur = dummy.next;

        while (cur != null && cur.next != null) {
            ListNode t = cur.next;
            pre.next = t;
            cur.next = t.next;
            t.next = cur;
            pre = cur;
            cur = pre.next;
        }

        return dummy.next;
    }
}
